WordPress/gutenberg
# Add comment id to all comments inside comments query loop

## What?

Adding the comment id to each comment on the Comments Query Loop.

## Why?

Feature parity with previous WordPress Comments markup before going to blocks. [ This WP Tavern post](https://wptavern.com/new-comment-related-blocks-arriving-with-wordpress-6-0#:~:text=The%20first%20issue%20I%20ran%20into%20was%20the%20missing%20comment%20ID%20in%20the%20front%2Dend%20output.) pointed out the issue with the new Comment Query Loop block:

> The first issue I ran into was the missing comment ID in the front-end output. This is necessary for the user’s browser to jump back to their comment after submitting one via the form. I also suspect this is necessary for the comment-reply JavaScript to work when clicking the reply link.

That suspicion turns out to be true:

| Before | After|
|-------|------|
| ![comment-before](https://user-images.githubusercontent.com/96308/163039569-ae4a4006-6f39-4e51-a3b5-b92873b82a0d.gif) | ![comment-after](https://user-images.githubusercontent.com/96308/163039872-cb18f52f-f161-4478-9b90-27a822d1154c.gif) |

Note how the comment form opens inline (in the position where the comment will show up, i.e. after the last child of the comment we're replying to). This is because of WordPress' [`comment-reply.js` script](https://github.com/WordPress/wordpress-develop/blob/2e4661c6c9193a1cd68cc2d113ef7ebea693375a/src/js/_enqueues/lib/comment-reply.js), which is [enqueued by the Post Comments Form block](https://github.com/WordPress/gutenberg/blob/4ed8a414d6314932046de99376ab49c98e129eb9/packages/block-library/src/post-comments-form/index.php#L31-L32).

Note that this behavior is consistent with a lot of (non-FSE) themes prior to Twenty Twenty-Two. Twenty Twenty-Two on the other hand uses the (monolithic) _Post Comments_ block which currently doesn't support this behavior; instead, the user always has to fill in the comment form at the bottom (below all other comments).

## How?
Just adding the comment id on the PHP function that is rendering.

## Concerns

There are a few problems with this: 
- [x] ~The comments form -- when rendered inline -- has some quirks in terms of layout (which we might be able to fix).~ _Done, see https://github.com/WordPress/gutenberg/pull/40268#issuecomment-1098391981_

- The comments form cannot currently be customized through the Site Editor. (This includes the UI in a broader sense, plus styling as well as copy such as "Reply to..." or "Cancel reply".) This is arguably not ideal, since it goes against the grain of enabling the user to use modular comments blocks to customize the way that comments are rendered. It could be argued that it's a result of the Post Comments Form block still being monolithic, and that having it is a necessary evil for the time being.
- `comment-reply.js` allows _some_ customization (e.g. of those strings) through `data-` attributes on the form. However, since these cannot be directly accessed through the Site Editor, we're left with the default values. This is a bit of a problem in terms of microtypography/styling: Note in the "After" gif that there is no whitespace between "Reply to admin" and "Cancel reply". The reason for this is that by default, there is no space in between those; themes either style them to add some margin, or customize the strings to add a space.
- Twenty TwentyTwo theme has an issue: "Cancel Reply" link is next to the username to reply to. This has been happening in all themes since Twenty Eleven, and it has been always solved by using CSS, which is an approach that we should not use in block themes. I've seen that this happens to most of Block themes on WordPress environment.
![Screenshot 2022-04-13 at 10 12 50](https://user-images.githubusercontent.com/37012961/163189579-997dfa88-2169-49c3-8f6e-caa8374b0eae.png)
cc: @WordPress/theme-team 

 
While there's an implicit connection between the Comments Query Loop and Post Comments Form block (the reply link won't work if the Post Comments Form block isn't also present), this isn't really due to this change: Even on `trunk`, if the Post Comments Form block was absent, the Comments Query Loop block's Reply links wouldn't work.

It's however arguable that the enqueuing of `comment-reply.js` should maybe happen in Comments Query Loop, rather than Post Comments Form.

We need to decide if we're okay with these glitches, or if we need to fix them. If the latter, we need to find out if they can be (easily) fixed (in time for WP 6.0).

## Testing Instructions
1.) Create a post with comments.
2.) Add Comment Query Loop.
3.) Add Post Comments Form.
3.) Check that ids are on the markup.
4.) Reply a comment and check that you end up in that reply after the page reload.
